File: ExamenUnidad2_PYTHON 1/ExamenUnidad2_PYTHON 1/ExamenUnidad2/Visitantes.py
from Persona import Persona

class Visitantes(Persona):
    IDENTIFICADOR = 0
    NUMERO_DE_VISITAS = 0

    
    def __init__(self, name=None, apellidos=None, fecha_de_nacimiento=None, CURP=None, fecha_de_registro=None, edad=None):
        if name is not None and apellidos is not None and fecha_de_nacimiento is not None and CURP is not None and fecha_de_registro is not None :
            super().__init__(name, apellidos, fecha_de_nacimiento, CURP)
            self.fecha_de_registro = fecha_de_registro
            self.num_de_visitas = 0
            self.esta_en_visita = False
            self.id = Visitantes.IDENTIFICADOR
            Visitantes.IDENTIFICADOR += 1
        else:
            # Aquí maneja el caso donde no se proporcionan todos los argumentos
            pass

    @staticmethod
    def get_NUMERO_DE_VISITAS():
        return Visitantes.NUMERO_DE_VISITAS

    @staticmethod
    def set_NUMERO_DE_VISITAS(NUMERO_DE_VISITAS):
        Visitantes.NUMERO_DE_VISITAS = NUMERO_DE_VISITAS

    def get_fecha_de_registro(self):
        return self.fecha_de_registro

    def set_fecha_de_registro(self, fecha_de_registro):
        self.fecha_de_registro = fecha_de_registro

    def get_num_de_visitas(self):
        return self.num_de_visitas

    def set_num_de_visitas(self):
        self.num_de_visitas += 1

    @staticmethod
    def get_IDENTIFICADOR():
        return Visitantes.IDENTIFICADOR

    @staticmethod
    def set_IDENTIFICADOR(IDENTIFICADOR):
        Visitantes.IDENTIFICADOR = IDENTIFICADOR

    def get_id(self):
        return self.id

    def set_id(self, id):
        self.id = id

    def is_esta_en_visita(self):
        return self.esta_en_visita

    def set_esta_en_visita(self, esta_en_visita):
        self.esta_en_visita = esta_en_visita

    def obtener_informacion_visitante(self):
        return (f"\nId: {self.id} \nNombre: {self.get_name()} \nEdad: {self.calcular_edad()} \nApellido: {self.get_apellidos()} \nFecha de nacimiento: {self.get_fecha_de_nacimiento()} \nCurp: {self.get_CURP()} \nFecha de registro: {self.fecha_de_registro} \nNúmero de visitas: {self.num_de_visitas}")

    def es_visitante(self, id, lista_visitantes):
        for visitante in lista_visitantes:
            if visitante.get_id() == id:

                return visitante
        return None

    def calcular_edad(self):
        # Calcular la edad en años
        # (Suponemos que la fecha de nacimiento está en formato "YYYY-MM-DD")
        nacimiento = self.get_fecha_de_nacimiento().split("-")
        anio_nacimiento = int(nacimiento[0])
        anio_actual = 2024  # Año actual (puede ser obtenido dinámicamente)
        return anio_actual - anio_nacimiento

    def obtener_descuento(self):
        # Cada 5 visitas, se aplica un descuento del 20%
        if self.num_de_visitas == 5:
            return 0.2
        else:
            return 0

File: ExamenUnidad2__JAVA/ExamenUnidad2/src/examenunidad2/Visitantes.java
package examenunidad2;
import java.util.ArrayList;

public class Visitantes extends Persona 
{
     public static int IDENTIFICADOR = 0;
    public static int NUMERO_DE_VISITAS = 0;
    private String fechaDeRegistro;
    private int numDeVisitas=0;
    private boolean estaEnVisita = false;
    private int id;

    public Visitantes() {
    }
    
    public Visitantes(String name, String apellidos,
            String fechaDeNacimiento, String CURP, String fechaDeRegistro) 
    {
        super(name, apellidos, fechaDeNacimiento, CURP);
        this.fechaDeRegistro = fechaDeRegistro;
        this.id = IDENTIFICADOR +=1;
        this.estaEnVisita=false;
    }

    public static int getNUMERO_DE_VISITAS() {
        return NUMERO_DE_VISITAS;
    }

    public static void setNUMERO_DE_VISITAS(int NUMERO_DE_VISITAS) {
        Visitantes.NUMERO_DE_VISITAS = NUMERO_DE_VISITAS;
    }

    public String getFechaDeRegistro() {
        return fechaDeRegistro;
    }

    public void setFechaDeRegistro(String fechaDeRegistro) {
        this.fechaDeRegistro = fechaDeRegistro;
    }

    public int getNumDeVisitas() {
        return numDeVisitas;
    }   

    public void setNumDeVisitas() {
        
        this.numDeVisitas++;
        
    }

    public static int getIDENTIFICADOR() {
        return IDENTIFICADOR;
    }

    public static void setIDENTIFICADOR(int IDENTIFICADOR) {
        Visitantes.IDENTIFICADOR = IDENTIFICADOR;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }
    
    
    
    public boolean getEstaEnVisita() {
        return estaEnVisita;
    }

    public void setEstaEnVisita(boolean estaEnVisita) {
        this.estaEnVisita = estaEnVisita;
    }
    
    public String obtenerInformacionVisitante()
    {
        return String.format("\nId: %d \nNombre: %s \nEdad: %d \nApellido: %s \nFecha de nacimiento: %s \nCurp: %s \nFecha de registro: %s \nNúmero de visitas: %d ",getId(), getName(), calcularEdad(), getApellidos(), getFechaDeNacimiento(), getCURP(),
             fechaDeRegistro, numDeVisitas);
    }

    
    public Visitantes esVisitante(int id, ArrayList <Visitantes> listavisitantes ){
        for(Visitantes visitante: listavisitantes){
            if(visitante.getId()==id){
                return visitante;
            }
        }
        return null;
    }
    public int calcularEdad() 
    {
        // Calcular la edad en años
        // (Suponemos que la fecha de nacimiento está en formato "YYYY-MM-DD")
        String[] nacimiento = fechaDeNacimiento.split("-");
        int anioNacimiento = Integer.parseInt(nacimiento[0]);
        int anioActual = 2024; // Año actual (puede ser obtenido dinámicamente)
        return anioActual - anioNacimiento;
    }
    public double obtenerDescuento() {
        // Cada 5 visitas, se aplica un descuento del 20%
        if (numDeVisitas == 5) {
            return 0.2;
        } else {
            return 0;
        }
    }
}
